overlord/ifacestate/apparmorprompting/prompting.go

```diff
@@ -0,0 +1,578 @@
+				break run_loop
+			}
+
+			// XXX: this debug log leaks information about internal activity
+			logger.Debugf("Got from kernel req chan: %v", req)
+			if err := m.handleListenerReq(req); err != nil {
+				// XXX: this log leaks information about internal activity
+				logger.Noticef("Error while handling request: %+v", err)
+			}
+		case <-m.tomb.Dying():
+			logger.Noticef("InterfacesRequestsManager tomb is dying, disconnecting")
+			break run_loop
+		}
+	}
+	return m.disconnect()
+}
+
+func (m *InterfacesRequestsManager) handleListenerReq(req *listener.Request) error {
+	userID := uint32(req.SubjectUID)
+	// TODO: immediately deny if req.SubjectUID() == 0 (root)
+	snap := req.Label // Default to apparmor label, in case process is not a snap
+	tag, err := naming.ParseSecurityTag(req.Label)
+	if err == nil {
+		// the triggering process is not a snap, so treat apparmor label as snap field
+		snap = tag.InstanceName()
+	}
+
+	// TODO: when we support interfaces beyond "home", do a proper selection here
+	iface := "home"
+
+	path := req.Path
+
+	permissions, err := prompting.AbstractPermissionsFromAppArmorPermissions(iface, req.Permission)
+	if err != nil {
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("error while parsing AppArmor permissions: %v", err)
+		response := listener.Response{
+			Allow:      false,
+			Permission: req.Permission,
+		}
+		return requestReply(req, &response)
+	}
+
+	remainingPerms := make([]string, 0, len(permissions))
+	satisfiedPerms := make([]string, 0, len(permissions))
+
+	// we're done with early checks, serious business starts now, and we can
+	// take the lock
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	matchedDenyRule := false
+	for _, perm := range permissions {
+		if yesNo, err := m.rules.IsPathAllowed(userID, snap, iface, path, perm); err == nil {
+			if !yesNo {
+				matchedDenyRule = true
+			} else {
+				satisfiedPerms = append(satisfiedPerms, perm)
+			}
+		} else {
+			if !errors.Is(err, requestrules.ErrNoMatchingRule) {
+				logger.Noticef("error while checking request against existing rules: %v", err)
+			}
+			// No matching rule found
+			remainingPerms = append(remainingPerms, perm)
+		}
+	}
+	if matchedDenyRule {
+		// XXX: this debug log leaks information about internal activity
+		logger.Debugf("request denied by existing rule: %+v", req)
+
+		// At least some permissions were allowed, but others were denied.
+		// Respond with this information by allowing all the permissions which
+		// were explicitly allowed, and let the kernel auto-deny the denied
+		// permissions, and any which not matched by allow rules.
+		responsePermissions, _ := prompting.AbstractPermissionsToAppArmorPermissions(iface, satisfiedPerms)
+		// Error should not occur, but if it does, responsePermissions are set
+		// to none, leaving it to AppArmor to default deny.
+		response := listener.Response{
+			Allow:      true,
+			Permission: responsePermissions,
```

Comment:
what if permission is empty? is that equivalent to deny? shouldn't we proceed to make a prompt then?

Reply:
If there's a matching deny rule (with higher precedence than any allow rule) for any permission, we immediately send back a denial for any permissions which aren't explicitly allowed. If we happen to have rules allowing some of the requested permissions, we might as well put those permissions in the allow field we're sending back to the kernel, but any explicitly denied permissions or permissions without a matching rule will be put in the deny field going back to the kernel.

Reply:
in listener we'll hit this snippet:
```
	perms, ok := response.Permission.(notify.FilePermission)
	...
		// allow permissions which kernel initially allowed, along with those
		// which the were initially denied but the user explicitly allowed.
		resp.Allow = msg.Allow | (uint32(perms) & msg.Deny)
		resp.Deny = 0
		resp.Error = 0
```

I'm wondering, shouldn't deny be filled with `~uint32(perms) & msg.Deny` instead of 0?

Reply:
would be good to clarify @bboozzoo question